### ISSUE TITLE
[Feature] Timezone support

### ISSUE THREAD

**Issue by pasztorl** [mariadb-operator/mariadb-operator#684]
When I create a MariaDB instance it uses UTC by default.
If I want to change the timezone, for example:

```
myCnf: |
  [mariadb]
  ...
  default_time_zone = 'Europe/Budapest'

```
The DB failed to start:
`[ERROR] Fatal error: Illegal or unknown default time zone ‘Europe/Budapest’
`

I see that by default the container have the `MYSQL_INITDB_SKIP_TZINFO=1` environment variable, so this explains the situation.

Question is:
- why it is set?
- how can I set the TZ in the "preferred" way?

Thanks

**Comment by mmontes11** [comment]
Hey there!

The initial reasoning behind this is that we were only willing to support the UTC timezone in the early iterations of the projects. I'm happy to drop the `MYSQL_INITDB_SKIP_TZINFO`  variable and instead set `UTC` as a default TZ in the default config file:
https://github.com/mariadb-operator/mariadb-operator/blob/d0c6f176987ffa722d07974ed4432328dcc6cc98/controller/mariadb_controller.go#L266

This can be overridden by the user using the `spec.myCnf` field.

Contributions and suggestions welcome!

**Comment by mbezhanov** [comment]
@pasztorl I've started working on a PR to implement what @mmontes11 suggested.

Until then, you can make the following modifications to your manifest file to achieve the result you desire:

```yaml
myCnf: |
  [mariadb]
  . . .
  default_time_zone = 'Europe/Budapest'

. . .

env:
  - name: MYSQL_INITDB_SKIP_TZINFO
    value: ""
```

Hope that helps!

**Comment by mmontes11** [comment]
Fantastic @mbezhanov ! Let me know if you need any help

**Comment by pasztorl** [comment]
@mbezhanov thank you very much!

**Comment by pasztorl** [comment]
@mbezhanov also mariadb image needs to be updated to load the tzdata to mysql db, because by default there is no TZ data on dbs, so db can not start if you change the TZ.

**Comment by mbezhanov** [comment]
@pasztorl can you elaborate, please?

The Docker entrypoint does this during initial startup:

```bash
if [ -z "$MARIADB_INITDB_SKIP_TZINFO" ]; then
  . . .
  mariadb-tzinfo-to-sql --skip-write-binlog /usr/share/zoneinfo \
    | docker_process_sql --dont-use-mysql-root-password --database=mysql
  . . .
fi
```

Zone info is taken from `/usr/share/zoneinfo`, which is already present in the image.

This only happens for fresh installations, though. Assuming you want to apply the changes I mentioned to an existing installation, a quick workaround would be to execute the following command in the container:

```bash
mysql_tzinfo_to_sql /usr/share/zoneinfo | mysql mysql -u root -p$MARIADB_ROOT_PASSWORD
```

You should be able to apply the updated manifest afterwards.

**Comment by pasztorl** [comment]
i see thanks!

**Comment by mmontes11** [comment]
This have been supported via https://github.com/mariadb-operator/mariadb-operator/pull/712! Thanks for your contribution @mbezhanov 👍🏻 

Closing!

**Comment by mmontes11** [comment]
This have been supported via https://github.com/mariadb-operator/mariadb-operator/pull/712! Thanks for your contribution @mbezhanov 👍🏻 

Closing!

**Comment by mmontes11** [comment]
We need to think better how to roll out this change. Updating the operator to the latest version breaks existing `MariaDB` databases provisioned with a previous version of the operator:

```bash
2024-07-18 15:31:30 0 [ERROR] Fatal error: Illegal or unknown default time zone 'UTC'
2024-07-18 15:31:30 0 [ERROR] Aborting
```

See:
- https://www.brucedev.com/posts/fatal-error-illegal-or-unknown-default-time-zone-in-mysql/

Will think about this and let you know

**Comment by mmontes11** [comment]
@mbezhanov No prob! Thanks!

I've made a couple of changes to avoid breaking existing `MariaDB` instances:
- TimeZones can be configured via the new `timeZone` field
- If `timeZone` is not provided, it defaults to `SYSTEM` and the timezone data is not loaded (`MYSQL_INITDB_SKIP_TZINFO = 1`)

See https://github.com/mariadb-operator/mariadb-operator/pull/726 for further details

@pasztorl to address your initial questions:

> why it is set? (`MYSQL_INITDB_SKIP_TZINFO`)

For performance reasons, see: https://github.com/MariaDB/mariadb-docker/issues/262#issuecomment-672375238. Not loading the timezone data  for better performance and defaulting the timezone to `SYSTEM` seems like good defaults for most cases.

> how can I set the TZ in the "preferred" way?

Easy, just set `timeZone = Europe/Budapest`  in your `MariaDB`.

This will be released in `v0.0.30`, stay tuned!
